File: tests/parity/ignore_store_probe.py
#!/usr/bin/env python3
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Actual adapter proposals committed through the candidate SQLite Store API."""
import argparse
import hashlib
import json
from pathlib import Path
import subprocess
import sys
import tempfile

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / 'engine/compat'))
from ignore_adapter import Adapter


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--candidate', type=Path, required=True)
    parser.add_argument('--output', type=Path, required=True)
    args = parser.parse_args()
    checks = []
    calls = []

    def lookup(kind, name, **kwargs):
        calls.append(name)
        return {'addresses': ['192.0.2.1'], 'error': None}

    adapters = {name: Adapter(dict(ignoreself=False, usedns='yes', allowipv6='no',
                                  ignoreip=['192.0.2.1'] if name == 'a' else []),
                              'original-store-profile', lookup=lookup) for name in ('a', 'b', 'c')}
    shared_name = 'dns:' + adapters['a'].shared_binding
    with tempfile.TemporaryDirectory(prefix='f2z-original-ignore-store-') as directory:
        root = Path(directory)
        request = root / 'request.json'

        def exchange(jail, **values):
            request.write_text(json.dumps(dict(jail=jail, shared_name=shared_name, **values)))
            process = subprocess.run([str(args.candidate.resolve()), str(root / 'state.db'),
                                      str(request)], capture_output=True, check=True)
            return json.loads(process.stdout)

        def stage(jail, hostname, current):
            return adapters[jail].stage(hostname, now=100,
                snapshot=json.loads(current['jail']['payload']) if current['jail']['payload'] else None,
                shared_snapshot=json.loads(current['shared']['payload']) if current['shared']['payload'] else None)

        def commit(jail, occurrence, proposal, current, *, fail=False, read_only=False):
            return exchange(jail, occurrence=occurrence,
                checkpoint=json.dumps(proposal['snapshot'], sort_keys=True),
                shared_payload=None if read_only else json.dumps(proposal['shared_snapshot'], sort_keys=True),
                jail_revision=current['jail']['revision'], shared_revision=current['shared']['revision'], fail=fail)

        empty = exchange('a')
        proposal = stage('a', 'original.test', empty)
        rejected = commit('a', 'a-1', proposal, empty, fail=True)
        assert rejected['outcome'] == 'InjectedFailure' and rejected['published'] is False
        assert rejected['jail'] == empty['jail'] and rejected['shared'] == empty['shared']
        assert rejected['cursor'] is None
        checks.append('injected-after-shared-write-rolls-back-jail-shared-and-cursor')
        retry = stage('a', 'original.test', exchange('a'))
        assert len(calls) == 2
        saved = commit('a', 'a-1', retry, empty)
        assert saved['outcome'] == 'committed' and saved['shared']['revision'] == 1 and saved['published'] is True
        assert retry['decisions'][0]['ignored'] is True
        checks.append('retry-recomputes-unpublished-dns-and-commits')
        other = exchange('b')
        cached = stage('b', 'original.test', other)
        assert len(calls) == 2 and cached['decisions'][0]['ignored'] is False
        assert cached['shared_snapshot'] == json.loads(other['shared']['payload'])
        saved_b = commit('b', 'b-1', cached, other, read_only=True)
        assert saved_b['outcome'] == 'committed' and saved_b['shared']['revision'] == 1
        checks.append('fresh-process-cross-jail-cache-restore-with-independent-ignore-state')
        checks.append('unchanged-shared-read-dependency-commits-without-revision-bump')
        stale = exchange('c')
        stale_proposal = stage('c', 'original.test', stale)
        next_a = exchange('a')
        fresh = stage('a', 'second.test', next_a)
        updated = commit('a', 'a-2', fresh, next_a)
        assert updated['shared']['revision'] == 2
        conflict = commit('c', 'c-1', stale_proposal, stale, read_only=True)
        assert conflict['outcome'] == 'StaleSharedCheckpoint' and conflict['published'] is False and conflict['ready'] is False
        assert conflict['jail']['revision'] == 0 and conflict['cursor'] is None
        assert conflict['shared'] == updated['shared']
        checks.append('stale-read-only-shared-dependency-rejects-entire-jail-transaction')
        current = exchange('c')
        refreshed = stage('c', 'original.test', current)
        accepted = commit('c', 'c-1', refreshed, current, read_only=True)
        assert accepted['outcome'] == 'committed' and accepted['shared']['revision'] == 2
        assert len(calls) == 3
        checks.append('refresh-and-retry-reuses-current-shared-dns')
    repository = Path(__file__).resolve().parents[2]
    files = ['engine/ignore_store_probe.zig', 'engine/core/record_store.zig',
             'engine/core/record_pipeline.zig', 'engine/compat/ignore_adapter.py', 'engine/compat/ignore_dns.py',
             'engine/compat/action_info.py', 'engine/compat/duration.py',
             'tests/parity/ignore_store_probe.py']
    receipt = dict(scope='candidate-only actual adapter→Pipeline→Store SQLite transaction; no daemon or matching claim',
                   checks=checks, passed=len(checks), lookup_trace=calls,
                   candidate_sha256=hashlib.sha256(args.candidate.read_bytes()).hexdigest(),
                   source_sha256={name: hashlib.sha256((repository / name).read_bytes()).hexdigest() for name in files})
    args.output.write_text(json.dumps(receipt, indent=2) + '\n')
    print(json.dumps(receipt))


if __name__ == '__main__':
    main()

File: engine/ignore_store_probe.zig
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Original fixture bridge to the actual durable record store, never actions.
const std = @import("std");
const durable = @import("core/record_store.zig");
const pipeline = @import("core/record_pipeline.zig");
const records = @import("core/source_record.zig");
const Request = struct {
    jail: []const u8,
    shared_name: []const u8,
    occurrence: ?[]const u8 = null,
    checkpoint: []const u8 = "",
    shared_payload: ?[]const u8 = null,
    jail_revision: u64 = 0,
    shared_revision: u64 = 0,
    fail: bool = false,
};
const Proposal = struct {
    request: Request,
    published: bool = false,
    fn restore(_: ?[]const u8, _: ?*anyopaque) !void {}
    fn publish(context: ?*anyopaque) void {
        const self: *Proposal = @ptrCast(@alignCast(context.?));
        self.published = true;
    }
    fn release(_: ?*anyopaque) void {}
    fn prepare(_: records.Record, context: ?*anyopaque) !pipeline.Prepared {
        const self: *Proposal = @ptrCast(@alignCast(context.?));
        return .{ .checkpoint = self.request.checkpoint, .disposition = "original-ignore-observation", .shared_state = .{ .name = self.request.shared_name, .expected_revision = self.request.shared_revision, .payload = self.request.shared_payload }, .context = self, .publish = publish, .release = release };
    }
};
pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const a = arena.allocator();
    const args = try std.process.argsAlloc(a);
    if (args.len != 3) return error.InvalidArguments;
    const bytes = try std.fs.cwd().readFileAlloc(a, args[2], 9 * 1024 * 1024);
    const parsed = try std.json.parseFromSlice(Request, a, bytes, .{});
    const request = parsed.value;
    var store = try durable.Store.open(a, args[1]);
    defer store.close();
    var proposal = Proposal{ .request = request };
    var owner = pipeline.Pipeline{ .store = &store, .jail = request.jail, .processor = .{ .context = &proposal, .prepare = Proposal.prepare, .restore = Proposal.restore } };
    var outcome: []const u8 = "read";
    if (request.occurrence) |occurrence| {
        if (request.fail) store.fail_at = .after_shared_checkpoint;
        var hash: [32]u8 = undefined;
        std.crypto.hash.sha2.Sha256.hash(occurrence, &hash, .{});
        try owner.restore(a);
        if (owner.revision != request.jail_revision) return error.InvalidFixtureRevision;
        outcome = "committed";
        pipeline.Pipeline.acknowledge(.{
            .source = "original-source",
            .occurrence = occurrence,
            .cursor = occurrence,
            .message = occurrence,
            .raw_hash = hash,
        }, &owner) catch |err| {
            outcome = @errorName(err);
        };
    }
    const jail = try store.snapshot(a, request.jail);
    const shared = try store.sharedSnapshot(a, request.shared_name);
    const cursor = try store.sourceCursor(a, request.jail, "original-source");
    try std.json.stringify(.{ .outcome = outcome, .jail = jail, .shared = shared, .cursor = cursor, .published = proposal.published, .ready = owner.ready }, .{}, std.io.getStdOut().writer());
}
